File: StackViewDemo/StackViewDemo/SCStackView.m
//
//  YXStackView.m
//  HaoFenShu
//
//  Created by SnowCheng on 16/5/27.
//  Copyright © 2016年 yunxiao. All rights reserved.
//

#import "SCStackView.h"

@interface SCStackView()

@property (nonatomic, strong) NSMutableArray *spaceViews;
@property (nonatomic, strong) NSMutableArray *layoutContentViews;

@end

@implementation SCStackView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self commonInit];
    }
    return self;
}

- (void)commonInit {
    self.direction = SCStackDirectionStyleHorizontal;
    self.style = SCStackLayoutStyleEqualWidth;
    self.layoutContentViews = [NSMutableArray array];
    self.spaceViews = [NSMutableArray array];
}

- (void)updateLayout {
    switch (self.containViews.count) {
        case 0:
            break;
        case 1: {
            [self addSubview:self.containViews[0]];
            [self.containViews[0] mas_makeConstraints:^(MASConstraintMaker *make) {
                make.left.equalTo(self).offset(self.leftMargin);
                make.right.equalTo(self).offset(self.rightMargin);
                make.top.equalTo(self).offset(self.topMargin);
                make.bottom.equalTo(self).offset(self.bottomMargin);
            }];
            break;
        }
        default: {
            if (self.style == SCStackLayoutStyleEqualMargin) {
                
                [self.layoutContentViews addObject:self.containViews[0]];
                
                for (int i = 1; i< self.containViews.count; i++) {
                    UIView *spaceView = [UIView new];
                    [self.layoutContentViews addObject:spaceView];
                    [self.layoutContentViews addObject:self.containViews[i]];
                    [self.spaceViews addObject:spaceView];
                }
            } else {
                self.layoutContentViews = [NSMutableArray arrayWithArray:self.containViews];
            }
            
            switch (self.direction) {
                case SCStackDirectionStyleHorizontal:
                    [self horzontalLayout];
                    break;
                case SCStackDirectionStyleVertical:
                    [self verticalLayout];
                    break;
            }
            
            break;
        }
    }
}

- (void)horzontalLayout {
    switch (self.style) {
        case SCStackLayoutStyleEqualMargin:
            [self horizontalEqualMarginLayout];
            break;
        case SCStackLayoutStyleEqualWidth:
            [self horizontalEqualWidthLayout];
            break;
    }
}

- (void)verticalLayout {
    switch (self.style) {
        case SCStackLayoutStyleEqualMargin:
            [self verticalEqualMarginLayout];
            break;
        case SCStackLayoutStyleEqualWidth:
            [self verticalEqualWidthLayout];
            break;
    }
}

- (void)horizontalEqualWidthLayout {
    for (int i = 0; i< self.layoutContentViews.count; i++) {
        [self addSubview:(self.layoutContentViews[i])];
        //公有
        [self.layoutContentViews[i] mas_makeConstraints:^(MASConstraintMaker *make) {
            make.top.equalTo(self).offset(self.topMargin);
            make.bottom.equalTo(self).offset(-self.bottomMargin);
        }];
        //第一个
        if (i == 0) {
            [self.layoutContentViews[i] mas_makeConstraints:^(MASConstraintMaker *make) {
                make.leading.equalTo(self).offset(self.leftMargin);
            }];
            continue;
        }
        //最后一个
        if (i == self.layoutContentViews.count - 1) {
            UIView *lastView = self.layoutContentViews[i-1];
            [self.layoutContentViews[i] mas_makeConstraints:^(MASConstraintMaker *make) {
                make.leading.equalTo(lastView.mas_trailing).offset(self.space);
                make.trailing.equalTo(self).offset(-self.rightMargin);
                make.width.equalTo(self.layoutContentViews[i-1]);
            }];
            continue;
        }
        UIView *lastView = self.layoutContentViews[i-1];
        [self.layoutContentViews[i] mas_makeConstraints:^(MASConstraintMaker *make) {
            make.leading.equalTo(lastView.mas_trailing).offset(self.space);
            make.width.equalTo(self.layoutContentViews[i-1]);
        }];
    }
}

- (void)horizontalEqualMarginLayout {
    for (int i = 0; i< self.layoutContentViews.count; i++) {
        [self addSubview:(self.layoutContentViews[i])];
        //公有
        [self.layoutContentViews[i] mas_makeConstraints:^(MASConstraintMaker *make) {
            make.top.equalTo(self).offset(self.topMargin);
            make.bottom.equalTo(self).offset(-self.bottomMargin);
        }];
        //第一个
        if (i == 0) {
            [self.layoutContentViews[i] mas_makeConstraints:^(MASConstraintMaker *make) {
                make.leading.equalTo(self).offset(self.leftMargin);
            }];
            continue;
        }
        //最后一个
        if (i == self.layoutContentViews.count - 1) {
            UIView *lastView = self.layoutContentViews[i-1];
            [self.layoutContentViews[i] mas_makeConstraints:^(MASConstraintMaker *make) {
                make.leading.equalTo(lastView.mas_trailing).offset(self.space);
                make.trailing.equalTo(self).offset(-self.rightMargin);
            }];
            continue;
        }
        UIView *lastView = self.layoutContentViews[i-1];
        [self.layoutContentViews[i] mas_makeConstraints:^(MASConstraintMaker *make) {
            make.leading.equalTo(lastView.mas_trailing).offset(self.space);
        }];
    }
    
    UIView *firstSpaceView = self.spaceViews[0];
    for (int i = 1; i< self.spaceViews.count; i++) {
        UIView *spaceView = self.spaceViews[i];
        [spaceView mas_makeConstraints:^(MASConstraintMaker *make) {
            make.width.equalTo(firstSpaceView);
        }];
    }
}

- (void)verticalEqualWidthLayout {
    for (int i = 0; i< self.layoutContentViews.count; i++) {
        [self addSubview:(self.layoutContentViews[i])];
        //公有
        [self.layoutContentViews[i] mas_makeConstraints:^(MASConstraintMaker *make) {
            make.left.equalTo(self).offset(self.leftMargin);
            make.right.equalTo(self).offset(-self.rightMargin);
        }];
        //第一个
        if (i == 0) {
            [self.layoutContentViews[i] mas_makeConstraints:^(MASConstraintMaker *make) {
                make.top.equalTo(self).offset(self.topMargin);
            }];
            continue;
        }
        //最后一个
        if (i == self.layoutContentViews.count - 1) {
            UIView *lastView = self.layoutContentViews[i-1];
            [self.layoutContentViews[i] mas_makeConstraints:^(MASConstraintMaker *make) {
                make.top.equalTo(lastView.mas_bottom).offset(self.space);
                make.bottom.equalTo(self).offset(-self.bottomMargin);
                make.height.equalTo(self.layoutContentViews[i-1]);
            }];
            continue;
        }
        UIView *lastView = self.layoutContentViews[i-1];
        [self.layoutContentViews[i] mas_makeConstraints:^(MASConstraintMaker *make) {
            make.top.equalTo(lastView.mas_bottom).offset(self.space);
            make.height.equalTo(self.layoutContentViews[i-1]);
        }];
    }
}

- (void)verticalEqualMarginLayout {
    for (int i = 0; i< self.layoutContentViews.count; i++) {
        [self addSubview:(self.layoutContentViews[i])];
        //公有
        [self.layoutContentViews[i] mas_makeConstraints:^(MASConstraintMaker *make) {
            make.left.equalTo(self).offset(self.leftMargin);
            make.right.equalTo(self).offset(-self.rightMargin);
        }];
        //第一个
        if (i == 0) {
            [self.layoutContentViews[i] mas_makeConstraints:^(MASConstraintMaker *make) {
                make.top.equalTo(self).offset(self.topMargin);
            }];
            continue;
        }
        //最后一个
        if (i == self.layoutContentViews.count - 1) {
            UIView *lastView = self.layoutContentViews[i-1];
            [self.layoutContentViews[i] mas_makeConstraints:^(MASConstraintMaker *make) {
                make.top.equalTo(lastView.mas_bottom).offset(self.space);
                make.bottom.equalTo(self).offset(-self.bottomMargin);
            }];
            continue;
        }
        UIView *lastView = self.layoutContentViews[i-1];
        [self.layoutContentViews[i] mas_makeConstraints:^(MASConstraintMaker *make) {
            make.top.equalTo(lastView.mas_bottom).offset(self.space);
        }];
    }
    
    UIView *firstSpaceView = self.spaceViews[0];
    for (int i = 1; i< self.spaceViews.count; i++) {
        UIView *spaceView = self.spaceViews[i];
        [spaceView mas_makeConstraints:^(MASConstraintMaker *make) {
            make.height.equalTo(firstSpaceView);
        }];
    }
}

- (void)setContainViews:(NSArray *)containViews {
    _containViews = containViews;
    [self updateLayout];
}

@end

File: SCStackView__Swift/SCStackView.swift
//
//  YXStackView.swift
//  PitureTextAttachment
//
//  Created by SnowCheng on 16/2/26.
//  Copyright © 2016年 SnowCheng. All rights reserved.
//

import UIKit

class SCStackView: UIView {

    enum DirectionStyle {
        case Horizontal
        case Vertical
    }
    
    enum LayoutStyle {
        case EqualWidth
        case EqualMargin
    }
    
// MARK: - Private property
    private var spaceViews = [UIView]()
    private var layoutContentViews = [UIView]()

// MARK: - Public  property
    var topMargin: CGFloat = 0
    var bottomMargin: CGFloat = 0
    var rightMargin: CGFloat = 0
    var leftMargin: CGFloat = 0
    var space: CGFloat = 0
    var direction: DirectionStyle = .Horizontal
    var style: LayoutStyle = .EqualWidth
    var containViews: [UIView] = [UIView](){
        didSet{
            for subview in subviews{
                subview.removeFromSuperview()
            }
            updateLayout()
        }
    }
    /// 边距 上 左 下 右
    var contentEdge: (top: CGFloat, left: CGFloat, bottom: CGFloat, right: CGFloat) = (0,0,0,0){
        didSet{
            topMargin = contentEdge.top
            leftMargin = contentEdge.left
            bottomMargin = contentEdge.bottom
            rightMargin = contentEdge.right
        }
    }

// MARK: - override method (Init)

// MARK: - Public method
    func updateLayout(){
        switch containViews.count {
        case 0:
            break
        case 1:
            addSubview(containViews[0])
            containViews[0].snp_makeConstraints(closure: { (make) in
                make.leading.equalTo(leftMargin)
                make.trailing.equalTo(-rightMargin)
                make.top.equalTo(topMargin)
                make.bottom.equalTo(-bottomMargin)
            })
        default:
            if style == .EqualMargin {
                layoutContentViews.append(containViews.first!)
                for i in 1..<containViews.count {
                    let spaceView = UIView()
                    layoutContentViews.append(spaceView)
                    layoutContentViews.append(containViews[i])
                    spaceViews.append(spaceView)
                }
            } else {
                layoutContentViews = containViews
            }
            
            switch direction {
            case .Horizontal:
                horizontalLayout()
            case .Vertical:
                verticalLayout()
            }
        }
       
    }

// MARK: - Private method
    private func horizontalLayout(){
        switch style {
        case .EqualMargin:
            horizontalEqualMarginLayout()
        case .EqualWidth:
            horizontalEqualWidthLayout()
        }
    }
    
    private func verticalLayout(){
        switch style {
        case .EqualMargin:
            verticalEqualMarginLayout()
        case .EqualWidth:
            verticalEqualWidthLayout()
        }
    }
    
    private func horizontalEqualWidthLayout(){
        for i in 0 ..< layoutContentViews.count{
            addSubview(layoutContentViews[i])
            //公有
            layoutContentViews[i].snp_makeConstraints { (make) -> Void in
                make.top.equalTo(self).offset(topMargin)
                make.bottom.equalTo(self).offset(-bottomMargin)
            }
            //第一个
            if i == 0{
                layoutContentViews[i].snp_makeConstraints { (make) -> Void in
                    make.leading.equalTo(self).offset(leftMargin)
                }
                continue
            }
            //最后一个
            if i == layoutContentViews.count - 1{
                layoutContentViews[i].snp_makeConstraints { (make) -> Void in
                    make.leading.equalTo(layoutContentViews[i-1].snp_trailing).offset(space)
                    make.trailing.equalTo(self).offset(-rightMargin)
                    make.width.equalTo(layoutContentViews[i-1])
                }
                continue
            }
            layoutContentViews[i].snp_makeConstraints { (make) -> Void in
                make.leading.equalTo(layoutContentViews[i-1].snp_trailing).offset(space)
                make.width.equalTo(layoutContentViews[i-1])
            }
        }
    }
    
    private func horizontalEqualMarginLayout(){
        for i in 0 ..< layoutContentViews.count{
            addSubview(layoutContentViews[i])
            //公有
            layoutContentViews[i].snp_makeConstraints { (make) -> Void in
                make.top.equalTo(self).offset(topMargin)
                make.bottom.equalTo(self).offset(-bottomMargin)
            }
            //第一个
            if i == 0{
                layoutContentViews[i].snp_makeConstraints { (make) -> Void in
                    make.leading.equalTo(self).offset(leftMargin)
                }
                continue
            }
            //最后一个
            if i == layoutContentViews.count - 1{
                layoutContentViews[i].snp_makeConstraints { (make) -> Void in
                    make.leading.equalTo(layoutContentViews[i-1].snp_trailing)
                    make.trailing.equalTo(self).offset(-rightMargin)
                }
                continue
            }
            layoutContentViews[i].snp_makeConstraints { (make) -> Void in
                make.leading.equalTo(layoutContentViews[i-1].snp_trailing)
            }
        }
        for view in spaceViews.dropFirst() {
            view.snp_makeConstraints(closure: { (make) in
                make.width.equalTo(spaceViews.first!)
            })
        }
    }
    
    private func verticalEqualWidthLayout(){
        for i in 0 ..< layoutContentViews.count{
            addSubview(layoutContentViews[i])
            //公有
            layoutContentViews[i].snp_makeConstraints(closure: { (make) in
                make.leading.equalTo(self).offset(leftMargin)
                make.trailing.equalTo(self).offset(-rightMargin)
            })
            //第一个
            if i == 0{
                layoutContentViews[i].snp_makeConstraints { (make) -> Void in
                    make.top.equalTo(self).offset(topMargin)
                }
                continue
            }
            //最后一个
            if i == layoutContentViews.count - 1{
                layoutContentViews[i].snp_makeConstraints { (make) -> Void in
                    make.top.equalTo(layoutContentViews[i-1].snp_bottom).offset(space)
                    make.bottom.equalTo(self).offset(-bottomMargin)
                    make.height.equalTo(layoutContentViews[i-1])
                }
                continue
            }
            layoutContentViews[i].snp_makeConstraints { (make) -> Void in
                make.top.equalTo(layoutContentViews[i-1].snp_bottom).offset(space)
                make.height.equalTo(layoutContentViews[i-1])
            }
        }
    }
    
    private func verticalEqualMarginLayout(){
        for i in 0 ..< layoutContentViews.count{
            addSubview(layoutContentViews[i])
            //公有
            layoutContentViews[i].snp_makeConstraints(closure: { (make) in
                make.leading.equalTo(self).offset(leftMargin)
                make.trailing.equalTo(self).offset(-rightMargin)
            })
            //第一个
            if i == 0{
                layoutContentViews[i].snp_makeConstraints { (make) -> Void in
                    make.top.equalTo(self).offset(topMargin)
                }
                continue
            }
            //最后一个
            if i == layoutContentViews.count - 1{
                layoutContentViews[i].snp_makeConstraints { (make) -> Void in
                    make.top.equalTo(layoutContentViews[i-1].snp_bottom).offset(space)
                    make.bottom.equalTo(self).offset(-bottomMargin)
                }
                continue
            }
            layoutContentViews[i].snp_makeConstraints { (make) -> Void in
                make.top.equalTo(layoutContentViews[i-1].snp_bottom).offset(space)
            }
        }
        
        for view in spaceViews.dropFirst() {
            view.snp_makeConstraints(closure: { (make) in
                make.height.equalTo(spaceViews.first!)
            })
        }
    }
    
    

// MARK: - @objc Action

// MARK: - lazy Loading
    

}
